app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -2,13 +2,196 @@
+        }
+        // TODO: do some other checking of input string
+    }
+
+    private URL buildRequestUrlWithCity(String city) throws MalformedURLException {
+        Uri builtUri = Uri.parse(getText(R.string.weather_api_entry_point).toString())
+                .buildUpon()
+                .appendQueryParameter("q", city)
+                .appendQueryParameter("appid", getText(R.string.weather_api_key).toString())
+                .appendQueryParameter("units", "metric")
+                .build();
+        return new URL(builtUri.toString());
+    }
+
+    private String parseTemperature(String response) throws JSONException, NullPointerException, IllegalArgumentException {
+        String temperature = null;
+
+        JSONObject json = new JSONObject(response);
+        JSONObject main  = json.getJSONObject("main");
+        temperature = main.getString("temp");
+        if (temperature == null) {
+            throw new NullPointerException("Couldn't parse response at method parseTempetature");
+        }
+        if (temperature.length() == 0) {
+            throw new IllegalArgumentException("Zero response length at method parseTempetature");
+        }
+        Log.d("myTag", "temperature = " + temperature);
+
+        return temperature;
+    }
+
+    private String doRequest(URL weatherEndpoint) throws IOException, NullPointerException, IllegalArgumentException {
+        Log.d("myTag", "doRequest start");
+        String respond = null;
+
+        BufferedReader reader = null;
+        InputStream stream = null;
+        HttpsURLConnection connection = null;
+
+        try {
+            connection = (HttpsURLConnection) weatherEndpoint.openConnection();
+            connection.setRequestProperty("User-Agent", getText(R.string.app_name).toString());
+            connection.setRequestMethod("GET");
+            connection.setReadTimeout(3000);
+            connection.connect();
+
+            if (connection.getResponseCode() == HttpsURLConnection.HTTP_OK) {
+                stream = connection.getInputStream();
+                reader = new BufferedReader(new InputStreamReader(stream));
+                StringBuilder buf = new StringBuilder();
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    buf.append(line).append("\n");
+                }
+                respond = buf.toString();
+                if (respond == null) {
+                    throw new NullPointerException("Got null String on respond in doRequest method.");
+                }
+                if (respond.length() == 0) {
+                    throw new IllegalArgumentException("Got zero-length String on respond in doRequest method.");
+                }
+            } else {
+                Log.d("myTag", "not 200");
+                throw new IOException("Respond code is: " + connection.getResponseCode());
+            }
```

Comment:
Do not throw NPE & IllegalArgumentException. 
if HTTP response code is not equal 200 (HTTP_OK) it doesn't mean that something is wrong on IO side. You can have 401 response which means that authentication is required to get a resource.

To simplify things right now, lets return null value in case of any error. 

Learn about RuntimeExceptions.